### ISSUE TITLE
MQTT_Connect is a blocking call

### ISSUE THREAD

**Issue by pietercardoen** [FreeRTOS/coreMQTT#183]
Dear

I am using coreMQTT in an event based embedded architecture and I faced problems with the function MQTT_Connect as it's a blocking function, waiting on the connack message. Due to this blocking behaviour, I never get to the point of processing a TCP receive event which is preventing the connack message to be received.

Why is this function implemented to be blocking?

PS. I currently patched the coreMQTT library to avoid this blocking behaviour. I am willing to share this.

**Comment by muneebahmed10** [comment]
Hi @pietercardoen,

I understand that you are referring to [`receiveConnack`](https://github.com/FreeRTOS/coreMQTT/blob/d79fbaccd2a3f2ee1a5b1ba922f2e0b1e0ef1f91/source/core_mqtt.c#L1823) in `MQTT_Connect`, and the call stack goes through [`receivePacket`](https://github.com/FreeRTOS/coreMQTT/blob/d79fbaccd2a3f2ee1a5b1ba922f2e0b1e0ef1f91/source/core_mqtt.c#L1535), [`recvExact`](https://github.com/FreeRTOS/coreMQTT/blob/d79fbaccd2a3f2ee1a5b1ba922f2e0b1e0ef1f91/source/core_mqtt.c#L880), and eventually [`recvFunc`](https://github.com/FreeRTOS/coreMQTT/blob/d79fbaccd2a3f2ee1a5b1ba922f2e0b1e0ef1f91/source/core_mqtt.c#L741), the receive function from the transport interface, where your application blocks. How have you implemented the transport interface? Can it be changed so that new incoming data can be read during the receive call? Could the TCP event be processed in an ISR so it will be ready during the Connect's receive call?

The coreMQTT API in core_mqtt.h is a managed MQTT API. The MQTT connection is not complete until the CONNACK has been received, so coreMQTT blocks for it. If there is no data available on the socket, it is expected for the scheduler to schedule another task for that duration so that coreMQTT will not waste any resources. If the managed API is not applicable to your use case, then there is also a low level serialization/deserialization API in [`core_mqtt_serializer.h`](https://github.com/FreeRTOS/coreMQTT/blob/d79fbaccd2a3f2ee1a5b1ba922f2e0b1e0ef1f91/source/include/core_mqtt_serializer.h) and documented [here](https://freertos.org/Documentation/api-ref/coreMQTT/docs/doxygen/output/html/mqtt_design.html#mqtt_serializers). The serialization API can serialize or deserialize all MQTT packets, and your application can take care of sending or receiving them over the network.

**Comment by abhidixi11** [comment]
Hello @pietercardoen please let us know if you have any further questions. Also if you could give details about your patch it would be great. Thanks.

**Comment by pietercardoen** [comment]
Dear abhidixi11

Apologies for my late reply. I was on holiday. My patch is quite simple: instead of waiting on the connack message, persist state in the context object. Reception of the connack message is handled in the mainloop.

From 4bbcfe7c9a68c04e6d8a50dc1346c343212ee0a1 Mon Sep 17 00:00:00 2001
From: Pieter Cardoen <pieter.cardoen@intulo.be>
Date: Tue, 7 Dec 2021 10:30:34 +0100
Subject: [PATCH] Change to make MQTT_Connect non-blocking

---
 source/core_mqtt.c                    | 147 +++++++++++++-------------
 source/include/core_mqtt.h            |   4 +
 source/include/core_mqtt_serializer.h |   3 +-
 3 files changed, 78 insertions(+), 76 deletions(-)

diff --git a/source/core_mqtt.c b/source/core_mqtt.c
index 8bbccfb..f0ce23d 100644
--- a/source/core_mqtt.c
+++ b/source/core_mqtt.c
@@ -27,6 +27,8 @@
 #include <string.h>
 #include <assert.h>
 
+#include <stdio.h>
+
 #include "core_mqtt.h"
 #include "core_mqtt_state.h"
 
@@ -261,7 +263,6 @@ static MQTTStatus_t sendPublish( MQTTContext_t * pContext,
  * #MQTTSuccess otherwise.
  */
 static MQTTStatus_t receiveConnack( const MQTTContext_t * pContext,
-                                    uint32_t timeoutMs,
                                     bool cleanSession,
                                     MQTTPacketInfo_t * pIncomingPacket,
                                     bool * pSessionPresent );
@@ -1307,8 +1308,54 @@ static MQTTStatus_t receiveSingleIteration( MQTTContext_t * pContext,
                                                   pContext->transportInterface.pNetworkContext,
                                                   &incomingPacket );
 
-    if( status == MQTTNoDataAvailable )
+    if(pContext->connectStatus == MQTTConnecting)
     {
+        // TODO
+        bool sessionPresent = false;
+
+        /* Read CONNACK from transport layer. */
+        if( status == MQTTSuccess )
+        {
+
+            status = receiveConnack( pContext,
+                                     pContext->cleanSession,
+                                     &incomingPacket,
+                                     &sessionPresent );
+        }
+
+        if( status == MQTTSuccess )
+        {
+            /* Resend PUBRELs when reestablishing a session, or clear records for new sessions. */
+            status = handleSessionResumption( pContext, sessionPresent );
+        }
+
+        if( status == MQTTSuccess )
+        {
+            LogInfo( ( "MQTT connection established with the broker." ) );
+            pContext->connectStatus = MQTTConnected;
+            /* Initialize keep-alive fields after a successful connection. */
+            pContext->keepAliveIntervalSec = pContext->keepAliveIntervalSec;
+            pContext->waitingForPingResp = false;
+            pContext->pingReqSendTimeMs = 0U;
+
+            pContext->appCallback( (MQTTContext_t *)pContext, &incomingPacket, NULL );
+        }
+        else
+        {
+            LogError( ( "MQTT connection failed with status = %s.",
+                        MQTT_Status_strerror( status ) ) );
+        }
+
+        if( status == MQTTNoDataAvailable ){
+            status = MQTTSuccess;
+        }
+
+        return status;
+    }
+
+    if( status == MQTTNoDataAvailable)
+    {
+
         if( manageKeepAlive == true )
         {
             /* Assign status so an error can be bubbled up to application,
@@ -1325,6 +1372,7 @@ static MQTTStatus_t receiveSingleIteration( MQTTContext_t * pContext,
     }
     else if( status != MQTTSuccess )
     {
+
         LogError( ( "Receiving incoming packet length failed. Status=%s",
                     MQTT_Status_strerror( status ) ) );
     }
@@ -1354,6 +1402,7 @@ static MQTTStatus_t receiveSingleIteration( MQTTContext_t * pContext,
 
     if( status == MQTTNoDataAvailable )
     {
+
         /* No data available is not an error. Reset to MQTTSuccess so the
          * return code will indicate success. */
         status = MQTTSuccess;
@@ -1459,16 +1508,13 @@ static MQTTStatus_t sendPublish( MQTTContext_t * pContext,
 /*-----------------------------------------------------------*/
 
 static MQTTStatus_t receiveConnack( const MQTTContext_t * pContext,
-                                    uint32_t timeoutMs,
                                     bool cleanSession,
                                     MQTTPacketInfo_t * pIncomingPacket,
                                     bool * pSessionPresent )
 {
     MQTTStatus_t status = MQTTSuccess;
     MQTTGetCurrentTimeFunc_t getTimeStamp = NULL;
-    uint32_t entryTimeMs = 0U, remainingTimeMs = 0U, timeTakenMs = 0U;
-    bool breakFromLoop = false;
-    uint16_t loopCount = 0U;
+    uint32_t remainingTimeMs = 0U, timeTakenMs = 0U;
 
     assert( pContext != NULL );
     assert( pIncomingPacket != NULL );
@@ -1476,53 +1522,21 @@ static MQTTStatus_t receiveConnack( const MQTTContext_t * pContext,
 
     getTimeStamp = pContext->getTime;
 
-    /* Get the entry time for the function. */
-    entryTimeMs = getTimeStamp();
+    /* Time taken in this function so far. */
+    timeTakenMs = calculateElapsedTime( getTimeStamp(), pContext->connectSendTimeMs );
 
-    do
+    if( timeTakenMs < pContext->connectTimeoutMs )
     {
-        /* Transport read for incoming CONNACK packet type and length.
-         * MQTT_GetIncomingPacketTypeAndLength is a blocking call and it is
-         * returned after a transport receive timeout, an error, or a successful
-         * receive of packet type and length. */
-        status = MQTT_GetIncomingPacketTypeAndLength( pContext->transportInterface.recv,
-                                                      pContext->transportInterface.pNetworkContext,
-                                                      pIncomingPacket );
-
-        /* The loop times out based on 2 conditions.
-         * 1. If timeoutMs is greater than 0:
-         *    Loop times out based on the timeout calculated by getTime()
-         *    function.
-         * 2. If timeoutMs is 0:
-         *    Loop times out based on the maximum number of retries config
-         *    MQTT_MAX_CONNACK_RECEIVE_RETRY_COUNT. This config will control
-         *    maximum the number of retry attempts to read the CONNACK packet.
-         *    A value of 0 for the config will try once to read CONNACK. */
-        if( timeoutMs > 0U )
-        {
-            breakFromLoop = ( calculateElapsedTime( getTimeStamp(), entryTimeMs ) >= timeoutMs ) ? true : false;
-        }
-        else
-        {
-            breakFromLoop = ( loopCount >= MQTT_MAX_CONNACK_RECEIVE_RETRY_COUNT ) ? true : false;
-            loopCount++;
-        }
-
-        /* Loop until there is data to read or if we have exceeded the timeout/retries. */
-    } while( ( status == MQTTNoDataAvailable ) && ( breakFromLoop == false ) );
+        /* Calculate remaining time for receiving the remainder of
+         * the packet. */
+        remainingTimeMs = pContext->connectTimeoutMs - timeTakenMs;
+    }
+    else{
+        status = MQTTConnectTimeout;
+    }
 
     if( status == MQTTSuccess )
     {
-        /* Time taken in this function so far. */
-        timeTakenMs = calculateElapsedTime( getTimeStamp(), entryTimeMs );
-
-        if( timeTakenMs < timeoutMs )
-        {
-            /* Calculate remaining time for receiving the remainder of
-             * the packet. */
-            remainingTimeMs = timeoutMs - timeTakenMs;
-        }
-
         /* Reading the remainder of the packet by transport recv.
          * Attempt to read once even if the timeout has expired.
          * Invoking receivePacket with remainingTime as 0 would attempt to
@@ -1548,6 +1562,8 @@ static MQTTStatus_t receiveConnack( const MQTTContext_t * pContext,
 
     if( status == MQTTSuccess )
     {
+//        uint16_t packetIdentifier;
+
         /* Update the packet info pointer to the buffer read. */
         pIncomingPacket->pRemainingData = pContext->networkBuffer.pBuffer;
 
@@ -1765,9 +1781,8 @@ MQTTStatus_t MQTT_Connect( MQTTContext_t * pContext,
     size_t remainingLength = 0UL, packetSize = 0UL;
     int32_t bytesSent;
     MQTTStatus_t status = MQTTSuccess;
-    MQTTPacketInfo_t incomingPacket = { 0 };
 
-    incomingPacket.type = ( uint8_t ) 0;
+    assert(pContext->getTime != NULL);
 
     if( ( pContext == NULL ) || ( pConnectInfo == NULL ) || ( pSessionPresent == NULL ) )
     {
@@ -1817,35 +1832,17 @@ MQTTStatus_t MQTT_Connect( MQTTContext_t * pContext,
         }
     }
 
-    /* Read CONNACK from transport layer. */
-    if( status == MQTTSuccess )
-    {
-        status = receiveConnack( pContext,
-                                 timeoutMs,
-                                 pConnectInfo->cleanSession,
-                                 &incomingPacket,
-                                 pSessionPresent );
-    }
-
-    if( status == MQTTSuccess )
-    {
-        /* Resend PUBRELs when reestablishing a session, or clear records for new sessions. */
-        status = handleSessionResumption( pContext, *pSessionPresent );
-    }
+    if( status == MQTTSuccess){
+        MQTTGetCurrentTimeFunc_t getTimeStamp = pContext->getTime;
 
-    if( status == MQTTSuccess )
-    {
-        LogInfo( ( "MQTT connection established with the broker." ) );
-        pContext->connectStatus = MQTTConnected;
-        /* Initialize keep-alive fields after a successful connection. */
+        /* Get the entry time for the function. */
+        pContext->connectStatus = MQTTConnecting;
+        pContext->connectSendTimeMs = getTimeStamp();
+        pContext->connectTimeoutMs = timeoutMs;
         pContext->keepAliveIntervalSec = pConnectInfo->keepAliveSeconds;
         pContext->waitingForPingResp = false;
         pContext->pingReqSendTimeMs = 0U;
-    }
-    else
-    {
-        LogError( ( "MQTT connection failed with status = %s.",
-                    MQTT_Status_strerror( status ) ) );
+        pContext->cleanSession = pConnectInfo->cleanSession;
     }
 
     return status;
diff --git a/source/include/core_mqtt.h b/source/include/core_mqtt.h
index 65de5bf..c5dc954 100644
--- a/source/include/core_mqtt.h
+++ b/source/include/core_mqtt.h
@@ -104,6 +104,7 @@ typedef void (* MQTTEventCallback_t )( struct MQTTContext * pContext,
 typedef enum MQTTConnectionStatus
 {
     MQTTNotConnected, /**< @brief MQTT Connection is inactive. */
+    MQTTConnecting,
     MQTTConnected     /**< @brief MQTT Connection is active. */
 } MQTTConnectionStatus_t;
 
@@ -221,7 +222,10 @@ typedef struct MQTTContext
     /* Keep alive members. */
     uint16_t keepAliveIntervalSec; /**< @brief Keep Alive interval. */
     uint32_t pingReqSendTimeMs;    /**< @brief Timestamp of the last sent PINGREQ. */
+    uint32_t connectSendTimeMs;    /**< @brief Timestamp of the last send CONNECT. */
+    uint32_t connectTimeoutMs;
     bool waitingForPingResp;       /**< @brief If the library is currently awaiting a PINGRESP. */
+    bool cleanSession;
 } MQTTContext_t;
 
 /**
diff --git a/source/include/core_mqtt_serializer.h b/source/include/core_mqtt_serializer.h
index bf3d974..1312fc7 100644
--- a/source/include/core_mqtt_serializer.h
+++ b/source/include/core_mqtt_serializer.h
@@ -105,7 +105,8 @@ typedef enum MQTTStatus
     MQTTNoDataAvailable, /**< No data available from the transport interface. */
     MQTTIllegalState,    /**< An illegal state in the state record. */
     MQTTStateCollision,  /**< A collision with an existing state record entry. */
-    MQTTKeepAliveTimeout /**< Timeout while waiting for PINGRESP. */
+    MQTTKeepAliveTimeout, /**< Timeout while waiting for PINGRESP. */
+    MQTTConnectTimeout
 } MQTTStatus_t;
 
 /**
-- 
2.32.0


Kind regards
Pieter Cardoen

**Comment by abhidixi11** [comment]
Thanks @pietercardoen .
If I understand correctly, by main loop, you mean handled by application ? It is still going to be the same behavior, i.e., the application is going to wait till CONN_ACK is received. So what is the purpose of making connect asynchronous ? As @cobusve mentioned earlier, you can't do anything else (like sending publish)  until you receive CONN_ACK. 

>I never get to the point of processing a TCP receive event which is preventing the connack message to be received.

I think we trying understand why you are facing the above problem.  You should able to implement your transport receive function in such way you can process TCP event and then process CONN_ACK.

Thanks.

**Comment by pietercardoen** [comment]
Dear abhidixi11

The purpose was indeed to obtain identical behaviour. It doesn't make sense to publish messages before a connack is received.

I was using this in an event based system where events of the TCP socket and events of the MQTT library are processed by one thread. As soon as MQTT_connect was started, this event loop got blocked, not allowing any other event to be processed until the MQTT_connect was finished.

TCP socket events needed to receive the connack message were queued but never handled, resulting in a blocking system.

Main reason for this problem is that I was using our existing coreMQTT/sparkplug implementation as backend for a QT graphical application. The problems I experienced was due to the event handling in Qt.

Kind regards
Pieter

**Comment by abhidixi11** [comment]
Thanks for clarification @pietercardoen . So basically you are using it in a single threaded system where TCP and MQTT and any other events are processed in a single thread. 

While (1)
{
MQTTSend() // Send is direct does not need event
TCPIPReceive() // check and process incoming TCP /IP /UDP packets the TCP/IP stack receive portion is implemented in this fucntion
MQTTReceive() // check and process incoming MQTT. 
}

So without calling TCPIPReceive() you wont's able to call MQTTReceive().

Our current design assumes TCP/IP packets will be processed in a separate thread.

**Comment by abhidixi11** [comment]
Thanks for confirming. We will consider your suggestion for future major release.  I'm closing the issue, please feel free to open a new one if you have any further questions. Thanks!
